src/Adapter/CartRule/Repository/CartRuleRepository.php

```diff
@@ -328,6 +329,76 @@ public function restrictCartRules(CartRuleId $cartRuleId, array $restrictedCartR
         );
     }
 
+    /**
+     * @param CartRuleId $cartRuleId
+     * @param CarrierId[] $restrictedCarrierIds
+     *
+     * @return void
+     */
+    public function setCarrierRestrictions(CartRuleId $cartRuleId, array $restrictedCarrierIds): void
+    {
+        $this->removeCarrierRestrictions($cartRuleId);
+
+        if (empty($restrictedCarrierIds)) {
+            return;
+        }
+
+        $checkedIds = [];
+        $insertValues = [];
+        foreach ($restrictedCarrierIds as $restrictedCarrierId) {
+            // skip duplicate ids if for some reason they exist
+            if (in_array($restrictedCarrierId, $checkedIds, true)) {
```

Comment:
Same comment as for country, if you want to check for duplicate values you should check for the actual int values, the the VO references